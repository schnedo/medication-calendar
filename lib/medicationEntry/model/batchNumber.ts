export default interface BatchNumber {
  readonly number: number;
}

const prefix = "Ch.-B.";
export function format(batchNumber: BatchNumber): string {
  return `${prefix} ${batchNumber.number}`;
}
