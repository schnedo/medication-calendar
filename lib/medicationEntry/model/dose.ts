export default interface Dose {
  readonly amount: number;
}

const unit = "ml";

export function format(dose: Dose): string {
  return `${dose.amount} ${unit}`;
}
