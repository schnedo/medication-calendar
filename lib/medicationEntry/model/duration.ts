export default interface Duration {
  minutes: number;
  hours: number;
}

function formatPart(number: number | undefined) {
  if (number === undefined) {
    return "00";
  }
  const raw = number.toString();
  return raw.length === 1 ? `0${raw}` : raw;
}

export function format(duration: Duration): string {
  return `${formatPart(duration.hours)}:${formatPart(duration.minutes)}`;
}
