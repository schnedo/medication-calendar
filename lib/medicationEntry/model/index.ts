export { default as BatchNumber } from "./batchNumber";
export { default as Dose } from "./dose";
export { default as Duration } from "./Duration";
export { default as Medicament } from "./medicament";
export { default as Medication } from "./medication";
export { default as MedicationEntry } from "./medicationEntry";

export * as Repository from "./repository";
