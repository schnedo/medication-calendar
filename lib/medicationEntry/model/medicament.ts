import BatchNumber from "./batchNumber";

export default interface Medicament {
  readonly name: string;
  readonly batchNumber: BatchNumber;
}
