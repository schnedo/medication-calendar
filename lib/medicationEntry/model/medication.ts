import Dose from "./dose";
import Medicament from "./medicament";
import { v4 } from "uuid";

export default class Medication {
  readonly id: string = v4();

  constructor(readonly dose: Dose, readonly medicament: Medicament) {}
}
