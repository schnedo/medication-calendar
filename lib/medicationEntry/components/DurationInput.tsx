import { ReactElement } from "react";
import {
  Box,
  FormControl,
  InputLabel,
  makeStyles,
  MenuItem,
  Select,
} from "@material-ui/core";
import { Duration } from "../model";
import { SelectInputProps } from "@material-ui/core/Select/SelectInput";

const range = (n: number): number[] => [...Array(n).keys()];

const createDurationMenuItem = (value: number) => (
  <MenuItem value={value} key={value}>
    {value}
  </MenuItem>
);

const useStyles = makeStyles((theme) => ({
  durationSelect: {
    display: "flex",
    "&>*": {
      marginRight: theme.spacing(1),
      "&:lastChild": {
        marginRight: theme.spacing(0),
      },
    },
  },
}));

interface NumberColumnProps {
  value: number;
  label: string;
  max: number;
  onChange: (value: number) => void;
}

function NumberColumn({
  value,
  label,
  max,
  onChange,
}: NumberColumnProps): ReactElement {
  const handleValueChange: SelectInputProps["onChange"] = (event) =>
    onChange(parseInt(event.target.value as string));

  const labelId = `${label}-label`;

  return (
    <FormControl>
      <InputLabel id={labelId}>{label}</InputLabel>
      <Select
        labelId={labelId}
        value={value.toString()}
        onChange={handleValueChange}
      >
        {range(max).map(createDurationMenuItem)}
      </Select>
    </FormControl>
  );
}

export interface DurationSelectProps {
  value: Duration;
  onChange: (duration: Duration) => void;
}

export default function DurationInput({
  value: { minutes, hours },
  onChange,
}: DurationSelectProps): ReactElement {
  const handleMinutesChange = (value: number) => {
    onChange({ minutes: value, hours });
  };
  const handleHoursChange = (value: number) => {
    onChange({ minutes, hours: value });
  };

  const { durationSelect } = useStyles();
  return (
    <Box className={durationSelect}>
      <NumberColumn
        label={"h"}
        value={hours ?? 0}
        max={23}
        onChange={handleHoursChange}
      />
      <NumberColumn
        value={minutes ?? 0}
        label={"min"}
        max={59}
        onChange={handleMinutesChange}
      />
    </Box>
  );
}
