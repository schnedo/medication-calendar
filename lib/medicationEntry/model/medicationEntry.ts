import Medication from "./medication";
import { BodyMass } from "../../contact/model";
import Duration from "./duration";
import { v4 } from "uuid";

export default class MedicationEntry {
  readonly id: string = v4();

  constructor(
    readonly date: Date,
    readonly medications: Medication[],
    readonly duration: Duration,
    readonly comments: string,
    readonly bodyMass: BodyMass,
  ) {}
}
