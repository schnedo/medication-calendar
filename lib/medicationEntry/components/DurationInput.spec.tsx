import { render } from "@testing-library/react";
import DurationInput from "./DurationInput";
import { Duration } from "../model";
import userEvent from "@testing-library/user-event";

describe.skip("DurationInput", () => {
  it("should render correctly", async () => {
    expect.hasAssertions();

    const duration: Duration = {
      minutes: 15,
      hours: 1,
    };
    const { container } = render(
      <DurationInput value={duration} onChange={() => undefined} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should call change handler with selected duration", async () => {
    expect.hasAssertions();

    const duration: Duration = {
      minutes: 15,
      hours: 1,
    };
    const changeHandler = jest.fn();
    const { getByRole, getByLabelText } = render(
      <DurationInput value={duration} onChange={changeHandler} />,
    );

    userEvent.click(getByLabelText("h"));
    userEvent.click(getByRole("option", { name: "2" }));
    expect(changeHandler).toHaveBeenCalledTimes(1);
    expect(changeHandler).toHaveBeenLastCalledWith({ minutes: 15, hours: 2 });
    userEvent.click(getByLabelText("min"));
    userEvent.click(getByRole("option", { name: "25" }));
    expect(changeHandler).toHaveBeenCalledTimes(2);
    expect(changeHandler).toHaveBeenLastCalledWith({ minutes: 25, hours: 1 });
  });
});
