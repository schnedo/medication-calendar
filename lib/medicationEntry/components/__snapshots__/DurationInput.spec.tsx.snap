// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationInput should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-2 makeStyles-durationSelect-1"
  >
    <div
      class="MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink="true"
        id="h-label"
      >
        h
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="h-label"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          role="button"
          tabindex="0"
        >
          1
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value="1"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink="true"
        id="min-label"
      >
        min
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="min-label"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          role="button"
          tabindex="0"
        >
          15
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value="15"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
